### PR TITLE
fix: refactored as-vending to use custom config instead of marshallSe…

### LINKED ISSUE #44: refactor use of utilities.MarshalSettings with use of custom configuration

### PR DESCRIPTION
**Pull request by marcpfuller**
refactored as-vending to use custom config instead of marshallSettings. This partly closes #44

Signed-off-by: Marc-Philippe Fuller <marc-philippe.fuller@intel.com>

### CODE REVIEW

File: as-vending/Dockerfile [hunk 1]
```diff
@@ -36,4 +36,4 @@ RUN apk --no-cache add zeromq
 COPY --from=builder /usr/local/bin/as-vending/res/configuration.toml /res/configuration.toml
 COPY --from=builder /usr/local/bin/as-vending/main /as-vending
 
-CMD [ "/as-vending", "-cp=consul.http://edgex-core-consul:8500", "-r"]
+CMD [ "/as-vending", "-cp=consul.http://edgex-core-consul:8500", "--registry", "--confdir=/res"]
```
Comment (lenny-intel) [inline as-vending/Dockerfile:39]: This change not needed. `-r` is short hand for `--registry` and the default for `confdir` is `./res`  which is valid.

File: as-vending/main.go [hunk 2]
```diff
@@ -5,50 +5,70 @@ package main
...
 	"github.com/edgexfoundry/app-functions-sdk-go/v2/pkg/transforms"
-	utilities "github.com/intel-iot-devkit/automated-checkout-utilities"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
 )
 
 const (
 	serviceKey = "as-vending"
 )
 
+type myApp struct {
```
Comment (lenny-intel) [inline as-vending/main.go:24]:
```suggestion
type vendingAppService struct {
```

File: as-vending/config/configuration.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+import (
+	"fmt"
+)
+
+type ServiceConfig struct {
+	AppCustom AppCustomConfig
+}
+
+type AppCustomConfig struct {
```
Comment (lenny-intel) [inline as-vending/config/configuration.go:11]:
```suggestion
type VendingConfig struct {
```

File: as-vending/config/configuration.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
+package config
+
+import (
+	"fmt"
+)
+
+type ServiceConfig struct {
+	AppCustom AppCustomConfig
```
Comment (lenny-intel) [inline as-vending/config/configuration.go:8]: AppCustiom was just a sample name for the example. Name needs to be specific in actual implementations

```suggestion
	Vending VendingConfig
```

File: as-vending/config/configuration.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		return false
+	}
+
+	*c = *configuration
+
+	return true
+}
+
+// Validate ensures your custom configuration has proper values.
+func (ac *AppCustomConfig) Validate() error {
```
Comment (lenny-intel) [inline as-vending/config/configuration.go:46]:
```suggestion
func (ac *VendingConfig) Validate() error {
```

File: as-vending/main.go [hunk 6]
```diff
@@ -5,50 +5,70 @@ package main
...
+	app.lc = app.service.LoggingClient()
 
-	// get the application settings from configuration.toml
-	appSettings := service.ApplicationSettings()
-	if appSettings == nil {
-		lc.Error("No application settings found")
-		os.Exit(1)
+	// retrieve the required configurations
+	app.serviceConfig = &config.ServiceConfig{}
+	if err := app.service.LoadCustomConfig(app.serviceConfig, "AppCustom"); err != nil {
```
Comment (lenny-intel) [inline as-vending/main.go:48]:
```suggestion
	if err := app.service.LoadCustomConfig(app.serviceConfig, "Vending"); err != nil {
```

File: as-vending/main.go [hunk 7]
```diff
@@ -5,50 +5,70 @@ package main
...
+		app.lc.Errorf("failed load custom configuration: %s", err.Error())
```
Comment (lenny-intel) [inline as-vending/main.go:49]:
```suggestion
		app.lc.Errorf("failed load custom Vending configuration: %s", err.Error())
```

File: as-vending/main.go [hunk 8]
```diff
@@ -5,50 +5,70 @@ package main
...
+		return 1
 	}
 
-	var vendingState functions.VendingState
-	vendingState.Configuration = new(functions.ServiceConfiguration)
-	vendingState.CommandClient = service.CommandClient()
-	if vendingState.CommandClient == nil {
-		lc.Error("Error command service missing from client's configuration")
-		os.Exit(1)
+	if err := app.serviceConfig.AppCustom.Validate(); err != nil {
```
Comment (lenny-intel) [inline as-vending/main.go:53]:
```suggestion
	if err := app.serviceConfig.Vending.Validate(); err != nil {
```

File: as-vending/main.go [hunk 9]
```diff
@@ -5,50 +5,70 @@ package main
...
+		app.lc.Errorf("failed to validate configuration: %v", err)
```
Comment (lenny-intel) [inline as-vending/main.go:54]:
```suggestion
		app.lc.Errorf("failed to validate Vending configuration: %v", err)
```

File: as-vending/main.go [hunk 10]
```diff
@@ -5,50 +5,70 @@ package main
...
+		return 1
 	}
 
-	// retrieve & parse the required application settings into a proper
-	// configuration struct
-	if err := utilities.MarshalSettings(appSettings, vendingState.Configuration, true); err != nil {
-		lc.Errorf("Application settings could not be processed: %v", err.Error())
-		os.Exit(1)
+	app.vendingState.Configuration = &app.serviceConfig.AppCustom
```
Comment (lenny-intel) [inline as-vending/main.go:58]:
```suggestion
	app.vendingState.Configuration = &app.serviceConfig.Vending
```

File: as-vending/main.go [hunk 11]
```diff
@@ -57,46 +77,48 @@ func main() {
...
 	// close event thread
-	vendingState.DoorClosedDuringCVWorkflow = false
-	vendingState.DoorCloseWaitThreadStopChannel = doorCloseStopChannel
+	app.vendingState.DoorClosedDuringCVWorkflow = false
+	app.vendingState.DoorCloseWaitThreadStopChannel = doorCloseStopChannel
 	// inference thread
-	vendingState.InferenceDataReceived = false
-	vendingState.InferenceWaitThreadStopChannel = inferenceStopChannel
+	app.vendingState.InferenceDataReceived = false
+	app.vendingState.InferenceWaitThreadStopChannel = inferenceStopChannel
```
Comment (brian-intel) [inline as-vending/main.go:94]: 80-94 would be a good candidate for a new vendingStateInitialize function to make the code a bit easier to read.

File: as-vending/res/configuration.toml [hunk 12]
```diff
@@ -57,7 +57,7 @@ Type="edgex-messagebus"
   Host = "localhost"
   Port = 59882
 
-[ApplicationSettings]
+[AppCustom]
```
Comment (lenny-intel) [inline as-vending/res/configuration.toml:60]:
```suggestion
[Vending]
```

File: as-vending/routes/controller_test.go [hunk 13]
```diff
@@ -165,7 +166,7 @@ func TestBoardStatus(t *testing.T) {
 		CVWorkflowStarted:              true,
 		DoorOpenWaitThreadStopChannel:  doorOpenStopChannel,
 		DoorCloseWaitThreadStopChannel: doorCloseStopChannel,
-		Configuration:                  new(functions.ServiceConfiguration),
+		Configuration:                  new(config.AppCustomConfig),
```
Comment (lenny-intel) [inline as-vending/routes/controller_test.go:169]:
```suggestion
		Configuration:                  new(config.VendingConfig),
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by jim-wang-intel — APPROVED** [review]
LGTM

**Review by lenny-intel — APPROVED** [review]
LGTM

**Review by brian-intel — APPROVED** [review]

File: as-vending/main.go (hunk 11 above)
Comment (marcpfuller) [reply as-vending/main.go:93]: moving this to an issue

**Pull request merged**